src/tools/wasm-ctor-eval.cpp

```diff
@@ -0,0 +1,445 @@
+  void setupEnvironment() {
+    // prepare scratch memory
+    stack.resize(STACK_SIZE);
+    // fill usable values for stack imports
+    auto total = STACK_START + STACK_SIZE;
+    globals["STACKTOP"] = Literal(int32_t(STACK_START));
+    globals["STACK_MAX"] = Literal(int32_t(STACK_START + STACK_SIZE));
+    // tell the module to accept writes up to the stack end
+    memorySize = total / Memory::kPageSize;
+  }
+
+  // flatten memory into a single segment
+  void flattenMemory() {
+    MemoryUtils::flatten(wasm.memory);
+  }
+};
+
+struct CtorEvalExternalInterface : EvallingModuleInstance::ExternalInterface {
+  Module* wasm;
+  EvallingModuleInstance* instance;
+
+  void init(Module& wasm_, EvallingModuleInstance& instance_) override {
+    wasm = &wasm_;
+    instance = &instance_;
+  }
+
+  void importGlobals(EvallingGlobalManager& globals, Module& wasm_) override {
+  }
+
+  Literal callImport(Import *import, LiteralList& arguments) override {
+    std::string extra;
+    if (import->module == "env" && import->base == "___cxa_atexit") {
+      extra = "\nrecommendation: build with -s NO_EXIT_RUNTIME=1 so that calls to atexit are not emitted";
+    }
+    throw FailToEvalException(std::string("call import: ") + import->module.str + "." + import->base.str + extra);
+  }
+
+  Literal callTable(Index index, LiteralList& arguments, WasmType result, EvallingModuleInstance& instance) override {
+    // we assume the table is not modified (hmm)
+    // look through the segments, try to find the function
+    for (auto& segment : wasm->table.segments) {
+      bool found = false;
+      Index start;
+      if (auto* c = segment.offset->dynCast<Const>()) {
+        start = c->value.getInteger();
+        found = true;
+      } else if (segment.offset->is<GetGlobal>()) {
+        // a segment loaded from a global is assumed to be loaded at 0. this is true
+        // in emscripten in the non-dynamic linking case
+        start = 0;
+        found = true;
+      }
+      if (found) {
+        auto end = start + segment.data.size();
+        if (start <= index && index < end) {
+          auto name = segment.data[index - start];
+          // if this is one of our functions, we can call it; if it was imported, fail
+          if (wasm->getFunctionOrNull(name)) {
+            return instance.callFunctionInternal(name, arguments);
+          } else {
+            throw FailToEvalException(std::string("callTable on imported function: ") + name.str);
+          }
+        }
+      }
+    }
+    throw FailToEvalException(std::string("callTable on index not found in static segments: ") + std::to_string(index));
+  }
+
+  Literal load(Load* load, Address addr) override {
+    // load from the memory segments
+    switch (load->type) {
+      case i32: {
+        switch (load->bytes) {
+          case 1: return load->signed_ ? Literal((int32_t)doLoad<int8_t>(addr)) : Literal((int32_t)doLoad<uint8_t>(addr));
```

Comment:
This big switch statement looks pretty nearly copy-pasted from https://github.com/WebAssembly/binaryen/blob/62e9f5d881e2d7e7f9f5da845ed2dbc176bc0bc5/src/shell-interface.h#L166 .
We can move the switch logic to a helper function in the parent class, that takes a template argument that we can swap out what we want to do with a given invoke.

So this could look like:
```c++
Literal load(Load* load, Address addr) override {
  switchOnLoad<doLoad>(load, addr);
}
```
and the one in `shell-interface.h` would be
```c++
template <typename T>
T getMemory(Address address) {
  return memory.get<T>(address);
}
Literal load(Load* load, Address addr) override {
  return switchOnLoad<getMemory>(load, addr);
}
```